File: java/business/src/main/java/business/services/UserService.java
package business.services;


import business.exceptions.AuthenticationException;
import business.utils.PasswordEncryptor;
import models.User;
import repository.interfaces.UserRepository;

public class UserService {
    private UserRepository userRepository;
    private PasswordEncryptor encryptor = new PasswordEncryptor();
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public void login(String username, String password) {
        var foundUserOptional = userRepository.findUserByUserName(username);

        var foundUser = foundUserOptional.orElseThrow(() -> {
            throw new AuthenticationException("Username doesn't exist!");
        });

        if (!encryptor.authenticate(password, foundUser.getPassword()))
            throw new AuthenticationException("Incorrect password!");
    }

    public void register(String username, String password) {
        if (userRepository.findUserByUserName(username).isPresent())
            throw new AuthenticationException("Username " + username + " is already used!");

        String hashPassword = encryptor.hash(password);
        User user = new User(username, hashPassword);
        userRepository.save(user);
    }
}

File: csharp/Business/Services/UserService.cs
using Business.Utils;
using Domain.Models;
using Repository.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Authentication;
using System.Text;
using System.Threading.Tasks;

namespace Business.Services
{
    public class UserService
    {
        private IUserRepository userRepository;
        private Encryptor encryptor = new Encryptor(); 

        public UserService(IUserRepository userRepository)
        {
            this.userRepository = userRepository;
        }

        public void LoginUser(String userName, String password)
        {
            var foundUser = userRepository.FindUserByUsername(userName);

            if (foundUser == null)
                throw new AuthenticationException("Username doesn't exist!");

            if (!encryptor.Authenticate(password, foundUser.Password))
                throw new AuthenticationException("Incorrect password!");
        }

        public void RegisterUser(String userName, String password)
        {
            if (userRepository.FindUserByUsername(userName) != null)
                throw new AuthenticationException("Username " + userName + " is already used!");

            String hashedPassword = encryptor.GetHashForPassword(password);
            User user = new User(userName, hashedPassword);
            userRepository.Add(user);
        }
    }
}
